### ISSUE TITLE
Camomile plugins export pd/libpd symbols

### ISSUE THREAD

**Issue by Spacechild1** [pierreguillot/Camomile#157]
The plugins should only export their entry points.

**Comment by pierreguillot** [comment]
Please, describe the problem and give further information.

**Comment by Spacechild1** [comment]
Ah, sorry, should have been more specific. The Windows binaries export all symbols from the statically linked libpd. Generally, a VST plugin should only export its entry points. 
![camomile_dependencies](https://user-images.githubusercontent.com/16126632/64162410-8ae47400-ce3f-11e9-887d-89467c714ea4.PNG)

**Comment by pierreguillot** [comment]
Ok, so it is a Windows-specific issue. I guess this is because the EXTERN macro is defined as `__declspec(dllexport) extern`. I should change that... 
Did you encounter any incompatibility?  
Perhaps this is a dumb question but as these symbols are exported, I guess the plugin could load externals... It would be great if you could try ;)

**Comment by Spacechild1** [comment]
> Did you encounter any incompatibility?

Don't really, but hiding symbols never hurts :-) It's more of an issue on Linux than on Windows.

>  I guess the plugin could load externals...

that's true! Do I have to compile externals with PDINSTANCE? On Windows, this should actually work, I might give it a try. I guess on Linux you would have to dlopen externals with RTLD_DEEPBIND in case someone else already exported the Pd symbols. I already use RTLD_DEEPBIND in [vstplugin~], because otherwise Camomile would take the symbols from the Pd executable instead of the statically linked libpd, so for loading externals you would have to do the same. Unless we say that loading Camomile in Pd is a bad idea in the first place, but actually it was Miller himself who needed this for a project.

**Comment by Spacechild1** [comment]
I think the fundamental problem is that Pd externals link against pd.dll and try to find/load a module of that name. At the moment I see two solutions:

1) Camomile links dynamically and ships libpd.dll (which externals for Camomile can link against)
2) for every Camomile plugin, you have to link the externals against the plugin (instead of pd.dll)

I tried 2) and at least it didn't crash but I don't think my external actually loaded. Unfortunately, Camomile doesn't give me any console printout, so it's hard to tell where it fails. When I have more time, I can change the code so I can see the error messages.

**Comment by pierreguillot** [comment]
> Camomile doesn't give me any console printout

The console is accessible by clicking on the top left flower. You can also define the warning level like in Pd.

**Comment by Spacechild1** [comment]
ha, cool! it can find my external but can't load it. Now I need to do some printf debugging to find the reason.

**Comment by pierreguillot** [comment]
Yes I think externals must be compiled with PDINSTANCE. I will check all the definitions used by Camomile. I still don’t have a windows machine so it’s difficult for me to work for windows specific issues :(
